File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        maxArea = 0

        def visit(i, j):
            if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] == 0:
                return 0

            grid[i][j] = 0

            return 1 + visit(i + 1, j) + visit(i - 1, j) + visit(i, j + 1) + visit(i, j - 1)
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    maxArea = max(visit(i, j), maxArea)
        
        return maxArea

File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int numIslands = 0;
        int m = grid.size();
        int n = grid[0].size();

        auto dfs = [&](this auto& self, int r, int c) -> void {
            if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == '0') {
                return;
            }

            grid[r][c] = '0';
            self(r + 1, c);
            self(r - 1, c);
            self(r, c + 1);
            self(r, c - 1);

            return;
        };

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == '1') {
                    ++numIslands;
                    dfs(i, j);
                }
            }
        }

        return numIslands;
    }
};
